src/utilities/data-formatter.ts

```diff
@@ -1,54 +1,65 @@
+
+/**TODO: might need to be extended to include string[][] */
+type Values = string[] | number[] | string;
+
 /**
  *
- * @param {String[]|Number[]} values
+ * @param {String|String[]|Number[]} values
  * @param maxThreshold
  * @constructor
  */
-class DataFormatter {
-  constructor(values, maxThreshold = 0) {
+export class DataFormatter {
+
+  indexTable: { [key: string]: number, [key: number]: number } = {};
+  characterTable: { [key: number]: string | number | null } = {};
+  characters: (string | number)[] = [];
+  specialIndexes: number[] = [];
+
+  constructor(private values?: Values | undefined, maxThreshold = 0) {
     if (values === undefined) return;
 
     this.values = values;
     // go over all characters and keep track of all unique ones seen
     // count up all characters
-    this.indexTable = {};
-    this.characterTable = {};
-    this.characters = [];
-    this.specialIndexes = [];
+
     this.buildCharactersFromIterable(values);
     this.buildTables(maxThreshold);
   }
 
-  buildCharactersFromIterable(values) {
-    const tempCharactersTable = {};
+  buildCharactersFromIterable(values: Values) {
```

Comment:
Each method needs a return type.

Reply:
It has the return type of "void" (will actually return undefined') automatically inferred by Typescript,  I can make it explicit:
` buildCharactersFromIterable(values: Values): void {`

If that is what we are doing everywhere.

Reply:
My IDE asks me to do this, I assumed it was part of the rules we have setup.

Reply:
Done